components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java

```diff
@@ -1496,4 +1482,67 @@ public boolean isDuplicateContextTemplate(String contextTemplate) throws APIMana
         result.put("isMore", isMore);
         return result;
     }
+
+    public boolean isAPIProductAvailable(APIProductIdentifier identifier) throws APIManagementException {
+        String path = APIConstants.API_ROOT_LOCATION + RegistryConstants.PATH_SEPARATOR +
+                identifier.getProviderName() + RegistryConstants.PATH_SEPARATOR +
+                identifier.getApiProductName() + RegistryConstants.PATH_SEPARATOR + identifier.getVersion();
+        try {
+            return registry.resourceExists(path);
+        } catch (RegistryException e) {
+            handleException("Failed to check availability of APIProduct :" + path, e);
+            return false;
+        }
+    }
+
+    public APIProduct getAPIProduct(APIProductIdentifier identifier) throws APIManagementException {
+        String apiProductPath = APIUtil.getAPIProductPath(identifier);
+        Registry registry;
+        try {
+            String apiTenantDomain = MultitenantUtils.getTenantDomain(
+                    APIUtil.replaceEmailDomainBack(identifier.getProviderName()));
+            int tenantId = ServiceReferenceHolder.getInstance().getRealmService().getTenantManager()
+                    .getTenantId(apiTenantDomain);
+            if (!MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(apiTenantDomain)) {
+                APIUtil.loadTenantRegistry(tenantId);
+            }
+
+            if (this.tenantDomain == null || !this.tenantDomain.equals(apiTenantDomain)) { //cross tenant scenario
+                registry = ServiceReferenceHolder.getInstance().getRegistryService().getGovernanceUserRegistry(
+                        MultitenantUtils.getTenantAwareUsername(
+                                APIUtil.replaceEmailDomainBack(identifier.getProviderName())), tenantId);
+            } else {
+                registry = this.registry;
+            }
+            GenericArtifactManager artifactManager = APIUtil.getArtifactManager(registry,
+                    APIConstants.API_PRODUCT_KEY);
+            Resource apiResource = registry.get(apiProductPath);
+            String artifactId = apiResource.getUUID();
+            if (artifactId == null) {
+                throw new APIManagementException("artifact id is null for : " + apiProductPath);
+            }
+            GenericArtifact apiArtifact = artifactManager.getGenericArtifact(artifactId);
+
+            APIProduct apiProduct = APIUtil.getAPIProduct(apiArtifact, registry);
+
+            //check for API visibility
+            if (APIConstants.API_GLOBAL_VISIBILITY.equals(apiProduct.getVisibility())) { //global api
+                return apiProduct;
+            }
+            if (this.tenantDomain == null || !this.tenantDomain.equals(apiTenantDomain)) {
+                throw new APIManagementException("User " + username + " does not have permission to view APIProduct : "
+                        + apiProduct.getId().getApiProductName());
+            }
+
+            return apiProduct;
```

Comment:
I think you can move `return apiProduct;` statement to at the end of the method and remove `return null;` in catch blocks.